src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/Reference.cs

```diff
@@ -266,18 +266,31 @@ public void LoadXml(XmlElement value)
                         // let the transform read the children of the transformElement for data
                         transform.LoadInnerXml(transformElement.ChildNodes);
                         // Hack! this is done to get around the lack of here() function support in XPath
-                        if (transform is XmlDsigEnvelopedSignatureTransform
-                            && _uri != null && (_uri.Length == 0 || _uri[0] == '#'))
+                        if (transform is XmlDsigEnvelopedSignatureTransform)
                         {
                             // Walk back to the Signature tag. Find the nearest signature ancestor
                             // Signature-->SignedInfo-->Reference-->Transforms-->Transform
                             XmlNode? signatureTag = transformElement.SelectSingleNode("ancestor::ds:Signature[1]", nsm);
 
                             // Resolve the reference to get starting point for position calculation.
-                            XmlNode? referenceTarget =
-                                _uri.Length == 0
-                                ? transformElement.OwnerDocument
-                                : SignedXml!.GetIdElement(transformElement.OwnerDocument, Utils.GetIdFromLocalUri(_uri, out bool _));
+                            // This needs to match the way CalculateSignature resolves URI references.
+                            XmlNode? referenceTarget = null;
+                            if (_uri == null || _uri.Length == 0)
```

Comment:
Previously, a null uri or a uri that doesn't start with `'#'` would skip all of this code.  

With this change, URIs that don't start with '#' will see `referenceTarget` as null, so there's no change there.. but now a null value for `_uri` matches OwnerDocument instead of nothing.  I can't think off of the top of my head what impact that will have, so can you explain why it was changed?

If the answer was "because otherwise [0] is hard", then the null check should just be put back into line 269:

```C#
if (transform is XmlDsigEnvelopedSignatureTransform && _uri != null)
```

Reply:
I guess what I got wrong is the comment... This should match how Reference.CalculateHashValue works. That case explicitly mentions URI null meaning the whole doc:

https://github.com/dotnet/runtime/blob/release/8.0/src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/Reference.cs#L375

Reply:
Ok, so after a bit of poking around I was able to craft a document with a valid enveloped signature that has a `<Reference>` element w/o any attributes (effectively causing it to have `Uri = null`):

```
<?xml version="1.0" encoding="UTF-8"?><hello><world>Hi</world><Signature xmlns="http://www.w3.org/2000/09/xmldsig#"><SignedInfo><CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#WithComments" /><SignatureMethod Algorithm="http://www.w3.org/2001/04/xmldsig-more#rsa-sha256" /><Reference><Transforms><Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#WithComments" /></Transforms><DigestMethod Algorithm="http://www.w3.org/2001/04/xmlenc#sha256" /><DigestValue>SVaCE5w9iLXTVYTKP1t/yjjmPXvWovMYpgljGgpgz2Y=</DigestValue></Reference></SignedInfo><SignatureValue>UHeu+oXIteU77QrnOUA10sZrJYHjX8BC4pF+mL9ha64XNmPL05qGbvXyNxZlERBaSxCqsLmHTmgEEIew/BtMn/6KvI4A2prs25/ZDRjMZ1MhXQ5tHJUM+0hwnyGEklAgFU2dYtbR8Gzj541/yWcaAyo71KXmkD6SeEDHCK5qwNM=</SignatureValue><KeyInfo><KeyValue><RSAKeyValue><Modulus>pIed5HbnnEgAyc84LYtvx8CDWUotUv5+skgVTay5KQzgbRQZRJFv3x3Uioaoyfcq0gG0MtO0ZATjKUJcGWTvJoqhvR6YjiM8iR5nnksXUYjCd+zqx7UuA3QXdShXt1JLOtEOv/WNnBM6IjpKnaRDBT1+bZwOst/UGrtXdQxCqXk=</Modulus><Exponent>AQAB</Exponent></RSAKeyValue></KeyValue></KeyInfo></Signature></hello>
```

Unfortunately, though, none of the built-in signature verification methods work out of the box on this document, neither in dotnet 6 nor 8. By default `XmlDsigEnvelopedSignatureTransform.GetOutput()` throws, e.g.:

```
An XmlDocument context is required for enveloped transforms.
   at System.Security.Cryptography.Xml.XmlDsigEnvelopedSignatureTransform.GetOutput()
   at System.Security.Cryptography.Xml.TransformChain.TransformToOctetStream(Object inputObject, Type inputType, XmlResolver resolver, String baseUri)
   at System.Security.Cryptography.Xml.Reference.CalculateHashValue(XmlDocument document, CanonicalXmlNodeList refList)
   at System.Security.Cryptography.Xml.SignedXml.CheckDigestedReferences()
   at System.Security.Cryptography.Xml.SignedXml.CheckSignature(AsymmetricAlgorithm key)
   at System.Security.Cryptography.Xml.SignedXml.CheckSignatureReturningKey(AsymmetricAlgorithm& signingKey)
   at System.Security.Cryptography.Xml.SignedXml.CheckSignature()
```

Other framework / online tools (e.g. paste the above XML into https://tools.chilkat.io/xmlDsigVerify.cshtml), successfully  validate that signature out of the box.

I only managed to validate the signature by explicitly loading the `XmlDocument` into the `Transform`, manually via `LoadInput`, before calling `CheckSignature`. In that case, `CheckSignature` returns `true` in dotnet 6, but it returns `false` in dotnet 8, basically exhibiting the same behavior as outlined in #95390, but for the `Uri == null` case:

```
    XmlDocument xmlDoc = new();
	xmlDoc.LoadXml(xmlStr.Dump());

	SignedXml signedXml = new(xmlDoc);
	var signature = xmlDoc.GetElementsByTagName("Signature", SignedXml.XmlDsigNamespaceUrl)[0] as XmlElement;
	signedXml.LoadXml(signature);
	
	// without this, CheckSignature throws
	((Reference)signedXml.SignedInfo.References[0]).TransformChain[0].LoadInput(xmlDoc);
	signedXml.CheckSignature().Dump();
```

Reply:
Code now has tests to show this is indeed the desired behavior in cases where there's a `<Reference>` element without attributes.